### ISSUE TITLE
[data grid] `scrollToIndexes` with lazyloading

### ISSUE THREAD

**Issue by m-fercher** [mui/mui-x#15810]
I created an example application based on the lazy loading documentation example here:
https://codesandbox.io/p/sandbox/confident-benji-37qqyl?file=%2Fsrc%2FDemo.js

After the first couple of datagrid entries are loaded from the server we want to use scrollToIndexes() to navigate to a specific rowIndex. We tried using apiRef.current.scrollToIndexes() with a timeout in a useEffect which does work but we want to directly react to the rows being loaded and navigate afterwards.

The navigation itself does work when using onStateChange but once the focus is reset (e.g. by scrolling using the scrollbar) the grid scrolls back to the previously used coordinates.

tldr; We want to scroll with lazyloading and refrain from using timeouts and the scroll is supposed to happen only once in the beginning. 

Also, I saw a similar issue here https://github.com/mui/mui-x/issues/15179 but we don't want to use timeouts as a solution.

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```

  System:
    OS: Windows 11 10.0.22631
  Binaries:
    Node: 18.20.4 - C:\Program Files\nodejs\node.EXE
    npm: 10.7.0 - C:\Program Files\nodejs\npm.CMD
    pnpm: Not Found
  Browsers:
    Chrome: Not Found
    Edge: Chromium (127.0.2651.74)
  npmPackages:
    @emotion/react: ^11.13.3 => 11.13.3
    @emotion/styled: ^11.13.0 => 11.13.0
    @mui/base: ^5.0.0-beta.4 => 5.0.0-beta.63
    @mui/core-downloads-tracker:  6.1.9
    @mui/icons-material: ^6.1.9 => 6.1.9
    @mui/lab: ^6.0.0-beta.17 => 6.0.0-beta.17
    @mui/material: ^6.0.0 => 6.1.9
    @mui/private-theming:  6.1.9
    @mui/styled-engine:  6.1.9
    @mui/system:  6.1.9
    @mui/types:  7.2.19
    @mui/utils:  6.1.9
    @mui/x-data-grid: 7.23.0 => 7.23.0
    @mui/x-data-grid-pro: ^7.23.0 => 7.23.0
    @mui/x-date-pickers: ^7.23.0 => 7.23.0
    @mui/x-internals:  7.23.0
    @mui/x-license: ^7.0.0 => 7.23.0
    @types/react: ^18.2.14 => 18.2.14
    react: ^18.2.0 => 18.2.0
    react-dom: ^18.2.0 => 18.2.0

```

</details>

Search keywords: scrollToIndexes, lazyLoading
Order ID: 102538

**Comment by michelengelen** [comment]
Interesting ... it is indeed very similar and maybe even a duplicate of #15179 
@arminmeh could you please have a look into that?

**Comment by m-fercher** [comment]
Hey @arminmeh. 

I just tried your recommended approach but it does not navigate to the selected coordinates:

https://codesandbox.io/p/sandbox/confident-benji-37qqyl

Is there anything I'm missing?
Thanks in advance

**Comment by arminmeh** [comment]
Hey @m-fercher 

It looks like that my example was wrong. Probably worked for me while hot reloading and having the rows there.
At the point where I have suggested to do the scrolling, the grid does not have the row ready, so it can't scroll to it yet.

One solution would be to use state change listener (like you did), but execute the scroll only once by using ref flag.

I will check a bit further and get back to you today or tomorrow.

**Comment by arminmeh** [comment]
@m-fercher 

Here is the [updated sandbox](https://codesandbox.io/p/sandbox/lazyloading-scroll-3lgtn3?file=%2Fsrc%2FDemo.js)

To avoid using `onStateChange` since it fires quite often, I have used `rowsSet` event.
This event is also fired multiple times before the skeleton rows are rendered and dimensions of the grid are stored in the state. Dimensions are needed for the scroll API to calculate the scroll position.

So, the subscription listens to the event until there is a dimension value in the state at which point ti scrolls to the desired position and unsubscribes itself from the event.

Let me know if this covers your requirements.

**Comment by m-fercher** [comment]
Hi @arminmeh, it's working now with the rowsSet event! Thank you so much.
